src/services/firm_quote_validator.ts

```diff
@@ -0,0 +1,147 @@
+import { BigNumber, RfqtFirmQuoteValidator, SignedOrder } from '@0x/asset-swapper';
+import { assetDataUtils, ERC20AssetData } from '@0x/order-utils';
+import { In } from 'typeorm';
+import { Repository } from 'typeorm/repository/Repository';
+
+import { ONE_SECOND_MS } from '../constants';
+import { MakerBalanceChainCacheEntity } from '../entities/MakerBalanceChainCacheEntity';
+import { logger } from '../logger';
+
+
+const BIG_NUMBER_ZERO = new BigNumber(0);
+const THRESHOLD_CACHE_EXPIRED_MS = 2 * 60 * ONE_SECOND_MS;
+
+
+export class PostgresBackedFirmQuoteValidator implements RfqtFirmQuoteValidator {
+    private readonly _chainCacheRepository: Repository<MakerBalanceChainCacheEntity>;
+    private readonly _cacheExpiryThresholdMs: number;
+
+    private static _calculateTakerFillableAmountsFromQuotes(quotes: SignedOrder[], makerLookup: {[key: string]: BigNumber}): {makerAddressesToAddToCache: string[]; takerFillableAmounts: BigNumber[]} {
+        const makerAddressesToAddToCacheSet: Set<string> = new Set();
+        const takerFillableAmounts =  quotes.map(quote => {
+            const makerTokenBalanceForMaker: BigNumber | undefined = makerLookup[quote.makerAddress];
+
+            // TODO: Add Prometheus hooks
+            if (makerTokenBalanceForMaker === undefined) {
+                makerAddressesToAddToCacheSet.add(quote.makerAddress);
+                return quote.takerAssetAmount;
+            }
+
+            // Order is fully fillable, because Maker has 100% of the assets
+            if (makerTokenBalanceForMaker.gte(quote.makerAssetAmount)) {
+                return quote.takerAssetAmount;
+            }
+
+            // Order is empty, return zero
+            if (quote.makerAssetAmount.lte(0)) {
```

Comment:
should this be `makerTokenBalanceForMaker.lte(0)` ?

Reply:
no this is checking that `makerAssetAmount` is greater than 0 because if not it will raise a "divide by 0" error on L41. If `makerTokenBalanceForMaker` is 0 then `partialFillableAmount` will gracefully become 0